src/Microsoft.Health.Extensions.DependencyInjection/TypeRegistrationBuilder.cs

```diff
@@ -122,9 +135,14 @@ public TypeRegistrationBuilder ReplaceService<T>()
     /// <exception cref="NotSupportedException">Throws when Type was not explicitly defined</exception>
     public TypeRegistrationBuilder AsImplementedInterfaces(Predicate<Type> interfaceFilter = null)
     {
-        Debug.Assert(
-            _firstRegisteredType != null || _registrationMode == TypeRegistration.RegistrationMode.Transient,
-            $"Using \"AsImplementedInterfaces()\" without calling \"AsSelf()\" first in registration for \"{_type.Name}\" could have inconsistent results.");
+        bool assertion = _firstRegisteredType != null || _registrationMode == TypeRegistration.RegistrationMode.Transient;
+        string msg = $"Using \"AsImplementedInterfaces()\" without calling \"AsSelf()\" first in registration for \"{_type.Name}\" could have inconsistent results.";
```

Comment:
Instead of warnings, should there be a setting on the `TypeRegistrationBuilder` when first created that enforces these assertions, like:
```csharp
IServiceCollection services = new ServiceCollection();
services
    .Add<T>(strictRegistration: true) // Or whatever name we like lol
    .AsSelf()
    .AsInterfaces();
```

Alternatively, this sort of detection would probably be a good candidate for a [custom Roslyn Analyzer](https://devblogs.microsoft.com/dotnet/how-to-write-a-roslyn-analyzer/)

Reply:
Instead of a Debug.Assert could we have it throw an exception? That way it would be still be caught by components downstream. I agree with will a strictMode or similar setting would be useful, but by default set it to be enabled. If there is a case where we know we need the "wrong" behavior we can override the exception check.